### PR TITLE
CLI tools refactoring

### PR DESCRIPTION
**Pull request by JustKappaMan**
Hi. Beautified and optimized some stuff without changing any logic and functionality.

### CODE REVIEW

**Comment by JustKappaMan** [comment]
@Madis0 It would be great to refactor `CurseForge to Packwiz-Modrinth.py` as well (using f-strings, pathlib, tomllib etc.). But there's too much "path magic" going on in there. So I didn't touch it to avoid breaking anything. Maybe you can do it as the author of the code.

File: CLI tools/CurseForge to Packwiz-Modrinth.py [hunk 1]
```diff
@@ -131,30 +134,38 @@ def main() -> int:
...
+                print(f"Moved {pack} to desktop")
         os.system(packwiz_exe_path + " refresh")
-    
-    if not refresh_only:
-        mmc_zip_root = str(Path(cf_zip_path).parents[0])
-        mmc_zip_path = mmc_zip_root + "\\Fabulously Optimized " + pack_version + ".zip"
-        #remove_mod_from_archive("Sodium", mmc_zip_path)
-        #remove_mod_from_archive("Iris", mmc_zip_path)
 
     return 0
```
Comment (osfanbuff63) [inline CLI tools/CurseForge to Packwiz-Modrinth.py:163]: Is this actually used anywhere?

**Review by osfanbuff63 — COMMENTED** [review]
Looks good to me in general, just one comment

File: CLI tools/CurseForge to Packwiz-Modrinth.py (hunk 1 above)
Comment (JustKappaMan) [reply CLI tools/CurseForge to Packwiz-Modrinth.py:163]: @osfanbuff63 I guess not. I removed all these lines since `mmc_zip_root` and `mmc_zip_path` are just created and not used. If they are -- it'd be better to just comment lines from 148 to 152 ofc.

**Comment by RozeFound** [comment]
Why remove type hints?

File: CLI tools/CurseForge to Packwiz-Modrinth.py (hunk 1 above)
Comment (osfanbuff63) [reply CLI tools/CurseForge to Packwiz-Modrinth.py:163]: I was actually talking about the `return 0` on line 154/163

**Comment by JustKappaMan** [comment]
> Why remove type hints?

Oops, my bad. Fixed.

File: CLI tools/CurseForge to Packwiz-Modrinth.py (hunk 1 above)
Comment (JustKappaMan) [reply CLI tools/CurseForge to Packwiz-Modrinth.py]: @osfanbuff63 oh, you're about `return 0`. Since it's default return code and nothing else can be explicitly returned by `main` -- just removed it as well.

File: CLI tools/CurseForge to Packwiz-Modrinth.py [hunk 2]
```diff
@@ -26,16 +26,18 @@
...
+    with ZipFile(from_zip, "r") as archive:
+        if from_file in archive.namelist():
+            print(f"Copying {from_desc} to {to_desc}")
+            archive.extract(from_file, to_path)
+            print(f"Copied {from_desc} to {to_desc}")
         else:
-            print("Skipped " + from_desc + " copying to " + to_desc + ", didn't exist")
+            print(f"Skipped {from_desc} copying to {to_desc}, didn't exist")
 
-def remove_from_archive(file_path: str, archive_path: str) -> None:
```
Comment (RozeFound) [inline CLI tools/CurseForge to Packwiz-Modrinth.py:37]: So why did you remove type hints?

File: CLI tools/CurseForge to Packwiz-Modrinth.py [hunk 3]
```diff
@@ -44,44 +46,45 @@ def remove_from_archive(file_path: str, archive_path: str) -> None:
                 files.append((zipinfo.filename, archive.read(zipinfo.filename)))
 
     with ZipFile(archive_path, "w") as archive:
-        for filename, content in files: archive.writestr(filename, content)
+        for filename, content in files:
+            archive.writestr(filename, content)
 
-def read_mod_meta(file_handle: BytesIO) -> dict:
```
Comment (RozeFound) [inline CLI tools/CurseForge to Packwiz-Modrinth.py:49]: Here as well

File: CLI tools/CurseForge to Packwiz-Modrinth.py [hunk 4]
```diff
@@ -44,44 +46,45 @@ def remove_from_archive(file_path: str, archive_path: str) -> None:
...
+            archive.writestr(filename, content)
 
-def read_mod_meta(file_handle: BytesIO) -> dict:
 
+def read_mod_meta(file_handle) -> dict:
     with ZipFile(file_handle) as archive:
         data = archive.read("fabric.mod.json")
-        return parse_json(data, strict=False)
-    
-def remove_mod_from_archive(mod_name: str, archive_path: str) -> None:
```
Comment (RozeFound) [inline CLI tools/CurseForge to Packwiz-Modrinth.py:55]: And here

File: CLI tools/CurseForge to Packwiz-Modrinth.py [hunk 5]
```diff
@@ -44,44 +46,45 @@ def remove_from_archive(file_path: str, archive_path: str) -> None:
...
                     return
 
-def main() -> int:
 
+def main():
     mod_files = os.listdir(mods_path)
     if not refresh_only:
         for item in mod_files:
-            os.remove( os.path.join(mods_path, item))
+            os.remove(os.path.join(mods_path, item))
```
Comment (RozeFound) [inline CLI tools/CurseForge to Packwiz-Modrinth.py:79]: Why not replace with pathlib functions?

**Review by RozeFound — CHANGES_REQUESTED** [review]

**Comment by RozeFound** [comment]
Other than that looks good, and I'm not asking you to write type hints, I simply want to understand why did you remove them.

**Comment by JustKappaMan** [comment]
> Other than that looks good, and I'm not asking you to write type hints, I simply want to understand why did you remove them.

Initially I decided to remove all return type hints since project is pretty small and not all functions were annotated in the first place.

> Why not replace with pathlib functions?

There's too much path-related code in this file. Don't want to break anything. Ofc it'd be better to completely rewrite it using pathlib. But I've no time for testing it.

**Comment by RozeFound** [comment]
> Initially I decided to remove all return type hints since project is pretty small and not all functions were annotated in the first place.

No matter how small the project is, writing type annotation (not only for return types, but also for arguments) is a good habit to develop.  So I would like you to bring back argument type hints, and, if you have time, to write it to other non annotated functions.

> There's too much path-related code in this file. Don't want to break anything.

Funny thing, if you read it closely, you would realize that a good chunk if this code is not used, and can be freely deleted.
Most of this code is kept "just in case", so the author could change the method exporting modpack to Modrinth,  which is I doubt would work after so much time of just gathering dust.

> Ofc it'd be better to completely rewrite it using pathlib. But I've no time for testing it.

You won't be able to test it anyway ¯\\_(ツ)\_/¯

**Comment by JustKappaMan** [comment]
> So I would like you to bring back argument type hints, and, if you have time, to write it to other non annotated functions.

@RozeFound I already did it in [962d099](https://github.com/Fabulously-Optimized/fabulously-optimized/pull/754/commits/962d099eb8fcc187c4b5d0207064771ea67db7b7)

**Comment by RozeFound** [comment]
> I already did it in [962d099](https://github.com/Fabulously-Optimized/fabulously-optimized/pull/754/commits/962d099eb8fcc187c4b5d0207064771ea67db7b7)

In [962d099](https://github.com/Fabulously-Optimized/fabulously-optimized/pull/754/commits/962d099eb8fcc187c4b5d0207064771ea67db7b7) you did only return type hints, but not function arguments

File: CLI tools/CurseForge to Packwiz-Modrinth.py [hunk 6]
```diff
@@ -93,30 +94,43 @@ def main():
...
+                    f'-i "{mmc_zip_path}"',
+                    "-f packwiz",
+                    "--modrinth-search loose",
+                    f'-o "{mmc_zip_root}"',
+                    f'-c "{packwiz_config}"',
+                    f"-v {pack_version}",
+                    "--provider-priority Modrinth CurseForge Other",
+                    "--scheme mmc_export_packwiz_output",
+                )
+            )
```
Comment (RozeFound) [inline CLI tools/CurseForge to Packwiz-Modrinth.py:132]: Did you know about [shlex.join](https://docs.python.org/3/library/shlex.html#shlex.join)?

File: CLI tools/CurseForge to Packwiz-Modrinth.py [hunk 7]
```diff
@@ -131,8 +145,20 @@ def main():
...
+                (
+                    "mmc-export",
+                    f'-i "{mmc_zip_path}"',
+                    "-f Modrinth",
+                    "--modrinth-search loose",
+                    f'-o "{mmc_zip_root}"',
+                    f'-c "{modrinth_config}"',
+                    f"-v {pack_version}",
+                    f'--scheme {"{name}-{version}"}',
+                )
```
Comment (RozeFound) [inline CLI tools/CurseForge to Packwiz-Modrinth.py:159]: Something like this:

```py
shlex.join((
    "mmc-export",
    "-i", mmc_zip_path,
    "-f", "Modrinth",
    "--modrinth-search", "loose",
    "-o", mmc_zip_root,
    "-c", modrinth_config,
    "-v", pack_version,
    "--scheme", "{name}-{version}"
))
```

**Review by RozeFound — APPROVED** [review]
Now this looks very good, I just mentioned that you don't use shlex when handling cmds.
Anyway approving, you've done a great job!

File: CLI tools/CurseForge to Packwiz-Modrinth.py (hunk 6 above)
Comment (JustKappaMan) [reply CLI tools/CurseForge to Packwiz-Modrinth.py:132]: Now i know. Thanks for the tip :slightly_smiling_face:

**Comment by RozeFound** [comment]
Anyway, @JustKappaMan, приятно видеть земляка в опенсорсе :) 
You don't plan on adding anything else?

**Review by osfanbuff63 — APPROVED** [review]
Assuming that we aren't changing `CurseForge to Packwiz-Modrinth.py` right now, LGTM

**Comment by JustKappaMan** [comment]
> You don't plan on adding anything else?

Nope :)

File: CLI tools/CurseForge to MultiMC-Git.py [hunk 8]
```diff
@@ -35,23 +38,23 @@ def copy_dir(from_path: Path, to_path: Path, from_desc: str, to_desc: str) -> No
...
 
 # CurseForge to MultiMC
-remove_dir(Path(mmc_path, "mods"), "MultiMC mods")
-remove_dir(Path(mmc_path, "config"), "MultiMC configs")
-remove_dir(Path(mmc_path, "resourcepacks"), "MultiMC resourcepacks")
-remove_file(Path(mmc_path, "options.txt"), "MultiMC options.txt")
-copy_dir(Path(cf_path, "mods"), Path(mmc_path, "mods"), "CurseForge mods", "MultiMC")
-copy_dir(Path(cf_path, "config"), Path(mmc_path, "config"), "CurseForge configs", "MultiMC")
-copy_dir(Path(cf_path, "resourcepacks"), Path(mmc_path, "resourcepacks"), "CurseForge resource packs", "MultiMC")
+remove_dir(mmc_path / "mods", "MultiMC mods")
```
Comment (Madis0) [inline CLI tools/CurseForge to MultiMC-Git.py:49]: If this has `/` then I think ln 9 should lack the last `/` in the path. Similar case for `cf_path` and `git_path` (ln 7).

File: CLI tools/CurseForge to Packwiz-Modrinth.py [hunk 9]
```diff
@@ -128,33 +145,51 @@ def main() -> int:
...
+                        "modrinth.index.json",
+                        git_path + "\\" + "Modrinth",
+                        "Modrinth manifest",
+                        "Git",
+                    )
                 os.replace(packwiz_path + "\\" + pack, os.path.expanduser("~/Desktop") + "\\" + pack)
-                print("Moved " + pack + " to desktop")
-        os.system(packwiz_exe_path + " refresh")
-    
-    if not refresh_only:
```
Comment (Madis0) [inline CLI tools/CurseForge to Packwiz-Modrinth.py:148]: The goal of this section is removing jars from the MMC zip, if CF decides to upload a mod and archive it in its site. While I haven't used it for a while, I may still need that in the future.

File: CLI tools/MultiMC to Git.py [hunk 10]
```diff
@@ -34,30 +36,32 @@ def copy_to_archive(from_path: Path, to_path: str, archive_path: Path) -> None:
 version = "0.0.0"
 pattern = re.compile(r"\d+\.\d+\.\d+-?\w*\.?\d*")
 
-with open(Path(mmc_path, "instance.cfg"), "r") as file:
+with open(mmc_path / "instance.cfg", "r") as file:
```
Comment (Madis0) [inline CLI tools/MultiMC to Git.py:39]: As in the first file, need to ensure that the path won't become `...//instance.cfg`

File: CLI tools/Update version.py [hunk 11]
```diff
@@ -2,12 +2,12 @@
 from pathlib import Path
 
 
-cf_path = Path(Path.home(), "curseforge/minecraft/Instances/Fabulously Optimized/")
-title_screen_path = Path(cf_path, "config/isxander-main-menu-credits.json")
-warning_path = Path(cf_path, "config/fabric_loader_dependencies.json")
+mmc_path = Path.home() / "curseforge/minecraft/Instances/Fabulously Optimized/"
+title_screen_path = mmc_path / "config/isxander-main-menu-credits.json"
+warning_path = mmc_path / "config/fabric_loader_dependencies.json"
```
Comment (Madis0) [inline CLI tools/Update version.py:7]: Same as previous comment

**Review by Madis0 — CHANGES_REQUESTED** [review]

File: CLI tools/CurseForge to MultiMC-Git.py (hunk 8 above)
Comment (JustKappaMan) [reply CLI tools/CurseForge to MultiMC-Git.py:49]: > If this has `/` then I think ln 9 should lack the last `/` in the path. Similar case for `cf_path` and `git_path` (ln 7).

`/` isn't about just adding `/`. It's overloaded `__div__` method of `Path` class with it's own functionality. There's no need to worry about possible incorrect paths when using `pathlib`. It handles everything automatically and won't add unneeded characters.

File: CLI tools/Update version.py (hunk 11 above)
Comment (JustKappaMan) [reply CLI tools/Update version.py:7]: > Same as previous comment

Same. See my reply to your first comment.

File: CLI tools/MultiMC to Git.py (hunk 10 above)
Comment (JustKappaMan) [reply CLI tools/MultiMC to Git.py:39]: > As in the first file, need to ensure that the path won't become `...//instance.cfg`

Same. See my reply to your first comment.

File: CLI tools/CurseForge to Packwiz-Modrinth.py (hunk 9 above)
Comment (JustKappaMan) [reply CLI tools/CurseForge to Packwiz-Modrinth.py:148]: > The goal of this section is removing jars from the MMC zip, if CF decides to upload a mod and archive it in its site. While I haven't used it for a while, I may still need that in the future.

OK, brought it back in [72d439b](https://github.com/Fabulously-Optimized/fabulously-optimized/pull/754/commits/72d439bd22d147aa2e4a880db7187ffab6e28ad2)

File: CLI tools/CurseForge to Packwiz-Modrinth.py [hunk 12]
```diff
@@ -90,32 +95,45 @@ def main() -> int:
...
     if mmc_export_packwiz_export and not refresh_only:
         mmc_zip_root = str(Path(cf_zip_path).parents[0])
         mmc_zip_path = mmc_zip_root + "\\Fabulously Optimized " + pack_version + ".zip"
         packwiz_config = git_path + "Packwiz\\mmc-export.toml"
 
-        cmd = f'mmc-export -i "{mmc_zip_path}" -f packwiz --modrinth-search loose -o "{mmc_zip_root}" -c "{packwiz_config}" -v {pack_version} --provider-priority Modrinth CurseForge Other --scheme mmc_export_packwiz_output'
-        os.system(cmd)
-
-        packwiz_zip_path = Path(mmc_zip_root) / "mmc_export_packwiz_output.zip" 
+        os.system(
```
Comment (Madis0) [inline CLI tools/CurseForge to Packwiz-Modrinth.py:120]: I think the command merging broke here, as I got the following error
> mmc-export: error: argument cmd: invalid choice: 'Optimized' (choose from 'gh-login', 'gh-logout', 'purge-cache')

and the rest failed. 

The command it joined was 

`"mmc-export -i 'c:\\Users\\Madis\\Desktop\\Fabulously Optimized 5.5.0-beta.4.zip' -f packwiz --modrinth-search loose -o 'c:\\Users\\Madis\\Desktop' -c 'C:\\Users\\Madis\\Documents\\GitHub\\fabulously-optimized\\Packwiz\\mmc-export.toml' -v 5.5.0-beta.4 --provider-priority Modrinth CurseForge Other --scheme mmc_export_packwiz_output"`

cc @RozeFound

**Review by Madis0 — CHANGES_REQUESTED** [review]

File: CLI tools/CurseForge to Packwiz-Modrinth.py (hunk 12 above)
Comment (RozeFound) [reply CLI tools/CurseForge to Packwiz-Modrinth.py:120]: How does in even brakes? It outputs basically the same thing as the old string-merging code 
shlex output: ```mmc-export -i 'c:\Users\Madis\Desktop\Fabulously Optimized 5.5.0-beta.4.zip' -f packwiz --modrinth-search loose -o 'c:\Users\Madis\Desktop' -c 'C:\Users\Madis\Documents\GitHub\fabulously-optimized\Packwiz\mmc-export.toml' -v 5.5.0-beta.4 --provider-priority Modrinth CurseForge Other --scheme mmc_export_packwiz_output```
old method output: ```mmc-export -i "c:\Users\Madis\Desktop\Fabulously Optimized 5.5.0-beta.4.zip" -f packwiz --modrinth-search loose -o "c:\Users\Madis\Desktop" -c "C:\Users\Madis\Documents\GitHub\fabulously-optimized\Packwiz\mmc-export.toml" -v 5.5.0-beta.4 --provider-priority Modrinth CurseForge Other --scheme mmc_export_packwiz_output```
The only difference is type of quotes

Comment (RozeFound) [reply CLI tools/CurseForge to Packwiz-Modrinth.py:120]: Oh, the type of quotes really does makes a difference on windows. My bad.

Comment (JustKappaMan) [reply CLI tools/CurseForge to Packwiz-Modrinth.py:120]: @RozeFound yep, it does. That's what Python docs say:
> Warning: The shlex module is only designed for Unix shells.
> The `quote()` function is not guaranteed to be correct on non-POSIX compliant shells or shells from other operating systems such as Windows.

Waiting for your fix or revert :slightly_smiling_face:

File: CLI tools/CurseForge to Packwiz-Modrinth.py [hunk 13]
```diff
@@ -90,32 +95,41 @@ def main() -> int:
...
+            "mmc-export",
+            "-i", mmc_zip_path,
+            "-f", "packwiz",
+            "--modrinth-search", "loose",
+            "-o", mmc_zip_root,
+            "-c", packwiz_config,
+            "-v", pack_version,
+            "--provider-priority", "Modrinth", "CurseForge", "Other",
+            "--scheme", "mmc_export_packwiz_output",
+        ); subprocess.call(args, shell=True)
```
Comment (osfanbuff63) [inline CLI tools/CurseForge to Packwiz-Modrinth.py:130]: Any particular reason we use `;` here?

File: CLI tools/CurseForge to Packwiz-Modrinth.py [hunk 14]
```diff
@@ -127,34 +141,54 @@ def main() -> int:
...
+        args = (
+            "mmc-export",
+            "-i", mmc_zip_path,
+            "-f", "Modrinth",
+            "--modrinth-search", "loose",
+            "-o", mmc_zip_root,
+            "-c", modrinth_config,
+            "-v", pack_version,
+            "--scheme", "{name}-{version}",
+        ); subprocess.call(args, shell=True)
```
Comment (osfanbuff63) [inline CLI tools/CurseForge to Packwiz-Modrinth.py:154]: See above

File: CLI tools/CurseForge to Packwiz-Modrinth.py (hunk 13 above)
Comment (RozeFound) [reply CLI tools/CurseForge to Packwiz-Modrinth.py:130]: You don't like it? It's for the sake of compactness mostly and grouping related things together

Comment (osfanbuff63) [reply CLI tools/CurseForge to Packwiz-Modrinth.py:130]: I mean, it makes some sense, but I was commenting on it since I don't usually see it used in anything that's not one line

Comment (RozeFound) [reply CLI tools/CurseForge to Packwiz-Modrinth.py:130]: And... What do you want from me exactly? I can remove it if you don't like how it looks

Comment (osfanbuff63) [reply CLI tools/CurseForge to Packwiz-Modrinth.py:130]: I thought there was something against using `;`, but it doesn't seem like it. I don't have the ability to resolve this though, so I guess this stays open until someone can resolve it.

File: CLI tools/CurseForge to Packwiz-Modrinth.py (hunk 14 above)
Comment (osfanbuff63) [reply CLI tools/CurseForge to Packwiz-Modrinth.py:154]: (also see above conversation for resolution)

File: CLI tools/CurseForge to Packwiz-Modrinth.py (hunk 13 above)
Comment (RozeFound) [reply CLI tools/CurseForge to Packwiz-Modrinth.py:130]: Well, the only thing against it is PEP8, since it's "not pythonic"
Direct quote from PEP8 docs: 
```Compound statements (multiple statements on the same line) are generally discouraged.```
So it's just a matter of style choice, I use semicolons when I see appropriate, that's all

**Comment by Madis0** [comment]
Worked for me.
